compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -1587,23 +1587,33 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             self.ty_to_value_string(rcvr_ty.peel_refs())
         };
         if let SelfSource::MethodCall(_) = source {
-            let first_arg = if let Some(CandidateSource::Impl(impl_did)) = static_candidates.get(0)
-                && let Some(assoc) = self.associated_value(*impl_did, item_name)
-                && assoc.kind == ty::AssocKind::Fn
-            {
+            let first_arg = static_candidates.get(0).and_then(|candidate_source| {
+                let (assoc_did, impl_ty) = match candidate_source {
+                    CandidateSource::Impl(impl_did) => {
+                        (*impl_did, self.tcx.type_of(*impl_did).instantiate_identity())
```

Comment:
Why don't we use the `rcvr_ty` for `CandidateSource::Impl`, too?

Reply:
given the following:
```rust
struct B<T> {
    _b: T
}
impl<T> B<T> {
    fn hello(_a: i32) {}
    fn test(_a: Self, _b: i32) {}
}

fn main() {
    let _b = B {_b: ""};
    _b.test(1);
}
```
`rcvr_ty`: `B<&str>`
`first_ty`: `B<T>`
`self.tcx.type_of(*impl_did).instantiate_identity()`: `B<T>`

Reply:
anything else here can improve?